File: web/src/actions/digest.js
/* Copyright (C) 2019 Monomax Software Pty Ltd
 *
 * This file is part of Dnote.
 *
 * Dnote is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Dnote is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Dnote.  If not, see <https://www.gnu.org/licenses/>.
 */

import * as digestsService from '../services/digests';
import { decryptDigest } from '../crypto/digest';

export const START_FETCHING = 'digest/START_FETCHING';
export const RECEIVE = 'digest/RECEIVE';
export const RECEIVE_ERROR = 'digest/RECEIVE_ERROR';

function receiveDigest(item) {
  return {
    type: RECEIVE,
    data: {
      item
    }
  };
}

function startFetchingDigestNotes() {
  return {
    type: START_FETCHING
  };
}

function receiveError(error) {
  return {
    type: RECEIVE_ERROR,
    data: {
      error
    }
  };
}

export function getDigest(cipherKeyBuf, digestUUID, demo) {
  return async dispatch => {
    try {
      dispatch(startFetchingDigestNotes());

      const digest = await digestsService.fetch(digestUUID, { demo });
      const digestDec = await decryptDigest(digest, cipherKeyBuf);
      dispatch(receiveDigest(digestDec));
    } catch (err) {
      console.log('Error fetching digest', err.stack);
      dispatch(receiveError(err.message));
    }
  };
}

File: pkg/server/api/presenters/digest.go
/* Copyright (C) 2019 Monomax Software Pty Ltd
 *
 * This file is part of Dnote.
 *
 * Dnote is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Dnote is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Dnote.  If not, see <https://www.gnu.org/licenses/>.
 */

package presenters

import (
	"time"

	"github.com/dnote/dnote/pkg/server/database"
)

// Digest is a presented digest
type Digest struct {
	UUID      string    `json:"uuid"`
	Notes     []Note    `json:"notes"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// PresentDigests presetns digests
func PresentDigests(digests []database.Digest) []Digest {
	ret := []Digest{}

	for _, digest := range digests {
		p := Digest{
			UUID:      digest.UUID,
			CreatedAt: digest.CreatedAt,
			UpdatedAt: digest.UpdatedAt,
		}

		ret = append(ret, p)
	}

	return ret
}

// PresentDigest presents a digest
func PresentDigest(digest database.Digest) Digest {
	ret := Digest{
		UUID:  digest.UUID,
		Notes: PresentNotes(digest.Notes),
	}

	return ret
}
